### DISCUSSION TITLE
Compiler error when using Visual Studio 2019

### DISCUSSION THREAD

**Discussion by blackparrot91** [nlohmann/json#3119] [Q&A]
Hi,

I am getting a repeated compiler error when trying to add the single_include header file to my project:

Severity	Code	Description	Project	File	Line	Suppression State
Error	C3861	'to_char_type': identifier not found 	myProject	C:\Git\myProject\json.h	13750	
Error	C3861	'to_char_type': identifier not found 	myProject	C:\Git\myProject\json.h	13757	
Error	C3861	'to_char_type': identifier not found 	myProject	C:\Git\myProject\json.h	13758	
Error	C3861	'to_char_type': identifier not found 	myProject	C:\Git\myProject\json.h	13775	
Error	C3861	'to_char_type': identifier not found 	myProject	C:\Git\myProject\json.h	13780

Compiler version: cl
Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30136 for x86

I have the following compiler command line switches (these are all Visual Studio defaults)
`/JMC /permissive- /Yu"pch.h" /ifcOutput "x64\Debug\" /GS /W3 /Zc:wchar_t /ZI /Gm- /Od /sdl /Fd"x64\Debug\vc142.pdb" /Zc:inline /fp:precise /D "_DEBUG" /D "MYPROJECT_EXPORTS" /D "_WINDOWS" /D "_USRDLL" /D "_WINDLL" /D "_UNICODE" /D "UNICODE" /errorReport:prompt /WX- /Zc:forScope /RTC1 /Gd /MDd /FC /Fa"x64\Debug\" /EHsc /nologo /Fo"x64\Debug\" /Fp"x64\Debug\myProject.pch" /diagnostics:column `

Adding the `permissive` flag results in a different error:
Severity	Code	Description	Project	File	Line	Suppression State
Error	C2947	expecting '>' to terminate template-parameter-list, found '*='	myProject	C:\Git\myProject\json.h	15258	
Error	C2988	unrecognizable template declaration/definition	myProject	C:\Git\myProject\json.h	15258	

Has anyone encountered this before? Any suggestions as to what I could try?

Many thanks in advance,
Mat

**Comment by blackparrot91** [comment]
I have managed to get it to compile by using the 'include' directory instead - so at this point my question is purely academic!

**Reply by nlohmann** [reply]
In order to asses your issue, we need the following information:

- What is the issue you have?

- Please describe the steps to reproduce the issue. Can you provide a small but working code example?

- What is the expected behavior?

- And what is the actual behavior instead?

- Which compiler and operating system are you using? Is it a [supported compiler](https://github.com/nlohmann/json#supported-compilers)?

- Did you use a released version of the library or the version from the `develop` branch?

- If you experience a compilation error: can you [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests)?
